File: src/Reader.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import org.apache.xmlrpc.XmlRpcException;

public class Reader extends Thread{
	
	public Client client;
	
	private BufferedReader buffReader = new BufferedReader(new InputStreamReader(System.in));
	private String inputText = "";
	
	public Reader(Client client){
		this.client = client;
	}

	@Override
	public void run() {
		//Infinite loop for listening to input characters
		System.out.println("== List of options == \n-join \n-signoff \n-start \n-bully  \n-host  \n-list  \n-exit");
		while(true){
			try{
				inputText = buffReader.readLine().trim();
				System.out.println("> " + inputText);
				selectedOption(inputText);
			}catch(IOException | XmlRpcException e){
				e.printStackTrace();
			}	
		}
	}
	
	private void selectedOption(String option) throws IOException, XmlRpcException{
			switch (inputText) {
				case "join":
					System.out.println("Enter IP:Port");
					String newNodeIP = buffReader.readLine().trim();
					client.join(newNodeIP);
					break;
				case "signoff":
					client.signOff();
					break;
				case "start":
					client.startConcatProcess();
					break;
				case "bully":
					client.startElection();
					break;
				case "host":
					System.out.println(Server.host + " is the current host");
					break;
				case "list":
					Client.showAllLists();
					break;
				case "exit":
					System.out.println("Quiting the program...");
					client.signOff();
					System.exit(0);
					break;
				default:
					System.err.println("Pardon, wrong input.");
					break;

			}
	}
}

File: c++/DSProject/Reader.cpp
#include "Reader.h"

Reader::Reader(Client* xmlRpcClient) : QThread(){

    connect(this,SIGNAL(join(QString)),xmlRpcClient,SLOT(join(QString)),Qt::UniqueConnection);
    connect(this,SIGNAL(signOff()),xmlRpcClient,SLOT(signOff()),Qt::UniqueConnection);
    connect(this,SIGNAL(startElection()),xmlRpcClient,SLOT(startElection()),Qt::UniqueConnection);
    connect(this,SIGNAL(startConcatProcess()),xmlRpcClient,SLOT(startConcatProcess()),Qt::UniqueConnection);
    connect(this,SIGNAL(stopOperations()),xmlRpcClient,SLOT(stopOperations()),Qt::UniqueConnection);
    connect(this,SIGNAL(runOverRpc(QString,QList<QVariant>)),xmlRpcClient,SLOT(runOverRpc(QString,QList<QVariant>)),Qt::UniqueConnection);
    connect(this,SIGNAL(showAllLists()),xmlRpcClient,SLOT(showAllLists()),Qt::UniqueConnection);
	connect(this, SIGNAL(echo(QString, QString)), xmlRpcClient, SLOT(echo(QString, QString)), Qt::UniqueConnection);
	connect(xmlRpcClient, SIGNAL(finishedTask()), this, SLOT(clientFinished()), Qt::UniqueConnection);

	busy = false;
}

void Reader::run(){
    //Infinite loop for listening to input characters
    std::cout << std::endl << "List of options: join, signoff, start, bully, host, list, echo, exit"<< std::endl;
	while (true){
        //if (!busy){
			std::string str;
			std::getline(std::cin, str);
			QString inputText(str.c_str());
			std::cout << std::endl << "> " << inputText.toStdString();
			selectedOption(inputText);
        }
        QCoreApplication::processEvents();
    //}
}

void Reader::selectedOption(QString option){
	busy = true;
	if (option == "join"){
        std::cout << std::endl << "Enter IP:Port"<< std::endl;
		std::string str;
        std::getline(std::cin, str);
        QString newNodeIP(str.c_str());
        emit join(newNodeIP);
	}
	else if (option == "signoff"){
		//System.out.println("Operation \"Sign Off\" initiated.");
        emit signOff();
	}
	else if (option == "start"){
        std::cout << std::endl << "concatenation process initiated.";
        emit startConcatProcess();
	}
	else if (option == "bully"){
        std::cout << std::endl << "Host election initiated.";
        emit startElection();
	}
	else if (option == "host"){
		std::cout << std::endl << glb::port << " is the current host";
		busy = false;
	}
	else if (option == "list"){
        emit showAllLists();
	}
    else if (option == "echo"){
        std::cout << std::endl << "Enter IP:Port"<< std::endl;
        std::string str;
        std::getline(std::cin, str);
        QString newNodeIP(str.c_str());
        std::cout << std::endl << "Enter echo msg"<< std::endl;
        std::string str2;
        std::getline(std::cin, str2);
        QString echoStr(str2.c_str());
        emit echo(newNodeIP,echoStr);
    }
	else if (option == "exit"){
        std::cout << std::endl << "Quiting the program...";
        emit signOff();
		QCoreApplication::exit();
	}
	else{
		std::cout << std::endl << "Pardon, wrong input." << std::endl;
		busy = false;
    }
}

void Reader::clientFinished(){
	busy = false;
}
